File: src/00022_GenerateParentheses.py
# backtrack
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def generate(left:int, right:int, S):
            if len(S) == n*2:
                return result.append(S)

            if left < n:
                generate(left+1, right, S + "(")
            
            if right < left:
                generate(left, right+1, S + ")")

            return
        
        result =[]
        generate(0, 0, "")

        return result

File: src/C#/00022_GenerateParentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        var result = new List<string>();
        Generate(0, 0, "", n, result);
        return result;
    }

    public void Generate(int open, int close, string input, int n, IList<string> result)
    {
        if (input.Length == n*2) 
        {
            result.Add(input);
        }

        if (open < n) 
        {
            Generate(open+1, close, input + "(", n, result);
        }

        if (close < open)
        {
            Generate(open, close+1, input + ")", n, result);
        }
    }
}